tools/codegen/api/autograd.py

```diff
@@ -0,0 +1,76 @@
+from dataclasses import dataclass
+from typing import Optional, Sequence, List, Tuple
+
+from tools.codegen.api.types import *
+from tools.codegen.model import *
+
+# Represents a saved variable invovled in backward calculation.
+# Note that it can be a derived property of an input argument, e.g.:
+# we could save `other.scalar_type()` instead of the entire `other` tensor.
+@dataclass(frozen=True)
+class Variable:
+    # Name of the saved variable.
+    # Suffix is appended if it's derived property, e.g.: `other_scalar_type`
+    name: str
+
+    # The cpp type string.
+    # TODO: change from raw string to model.Type
+    type: str
+
+    # The expression to read the derived property at save time, e.g.:
+    # `other.scalar_type()`.
+    expr: str
+
+# Represents a backward formula that calculates derivatives for one
+# or more tensors.
+@dataclass(frozen=True)
+class Derivative:
+    # The formula string (legit C++ expression).
+    # Note that expressions against input arguments are replaced with the
+    # corresponding saved variables.
+    # E.g.: `mul_tensor_backward(grad, self, other.scalar_type())`
+    #    -> `mul_tensor_backward(grad, self, other_scalar_type)`
+    formula: str
+
+    # Names of the arguments for which this formula calculates derivatives.
+    var_names: Tuple[str, ...]
+
+    # Saved inputs that are referenced by the formula.
+    saved_inputs: Tuple[Variable, ...]
+
+    # Saved outputs that are referenced by the formula.
+    saved_outputs: Tuple[Variable, ...]
+
+# Represents differentiability info for a NativeFunction.
+@dataclass(frozen=True)
+class DifferentiabilityInfo:
+    # The base name read from derivatives.yaml.
+    name: str
+
+    # The matching native function.
```

Comment:
nit: can we add a comment there to say if it is the exact same overload or the base version?

Reply:
Added inline comments:
```
    # There can be multiple NativeFunction having the same base name:
    #  - different overloads with different types of input arguments;
    #  - in-place/out/functional variants of the same function;
    #
    # We first use the schema string (under the 'name' key) in derivatives.yaml
    # to find the NativeFunction having the same schema string.
    # Then we find the in-place/out/functional variants of the matching function.
    # Among these variants, we choose the one having the same name as the
    # derivatives.yaml entry. If there is no exact match, then we choose the
    # in-place variant, if any.
    # TODO: maybe the logic to search for all variants is no longer necessary?
```

After taking a closer look, I actually feel the `canonical_function()` doesn't seem to be necessary:
In the first step, we already find the function with the exact same function schema string - there
shouldn't be a better match among in-place/out/functional variants.